### ISSUE TITLE
[Bug]: All versions of Electron (chromium upstream bug) fail to open on Ubuntu 24.04+ due to kernel.apparmor_restrict_unprivileged_userns=1 by default

### ISSUE THREAD

**Issue by theofficialgman** [electron/electron#41066]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

All

### What operating system are you using?

Ubuntu

### Operating System Version

Ubuntu 24.04

### What arch are you using?

x64

### Last Known Working Electron version

None

### Expected Behavior

Applications built using electron can launch

### Actual Behavior

Applications cannot launch unless they each ship an apparmor override

### Additional Information

added optionally in Ubuntu 23.10, enabled by default and going forward in Ubuntu 24.04 LTS
https://ubuntu.com/blog/ubuntu-23-10-restricted-unprivileged-user-namespaces

There is a (not public) chromium bug report for this apparently. Maybe electron developers can see and track its progress 
https://bugs.chromium.org/p/chromium/issues/detail?id=1459821
https://bugs.launchpad.net/ubuntu/+source/linux-meta-nvidia-5.19/+bug/2017980/comments/27
(ignore the package the bug was created under in ubuntu, this has nothing to do with nvidia and affects all kernels)

**Comment by LeviMarvin** [comment]
I have the same problem. All apps based on Electron will not be able to run on Ubuntu 24.04 LTS. Tested Motrix and QQ.

If running the app with sudo (also `--no-sandbox`), then will get a blank window with nothing.

**Comment by LeviMarvin** [comment]
The temporary solution is disabling the kernel.apparmor_restrict_unprivileged_userns:

```
sudo sysctl -w kernel.apparmor_restrict_unprivileged_userns=0
```

**Comment by MikeMcC399** [comment]
The release notes [Ubuntu 24.04 Release notes > Unprivileged user namespace restrictions](https://discourse.ubuntu.com/t/noble-numbat-release-notes/39890#unprivileged-user-namespace-restrictions-15) refer to the breaking change in this [release](https://ubuntu.com/blog/canonical-releases-ubuntu-24-04-noble-numbat) on Apr 25, 2024.

**Comment by MennoLabs** [comment]
For Balena Etcher, starting it with --no-sandbox --disable-gpu-sandbox fixes it.
Just make an alias or a script.

**Comment by yvonxiao** [comment]
This issue is caused by `apparmor`'s default policy, which lacks a specific profile in the `/etc/apparmor.d/` directory. Currently, the workaround is to manually add the profile. you can refer to the latest version of the code for guidance.It's not necessary to disable sandbox or modify kernel parameters to resolve this.

**Comment by ravicious** [comment]
@yvonxiao Could you share links to what exactly the VSCode team did to take care of this? I searched recent pull requests for `apparmor`, `profile`, and `wayland`, but I couldn't find anything relevant.

The only thing is a comment talking about changing `code.desktop` https://github.com/microsoft/vscode/issues/212430#issuecomment-2109960783 but as far as I understand it's about forcing Wayland + adding some extra flags rather than manually adding some kind of profile?

**Comment by floviolleau** [comment]
Hi,

I did something like this for mattermost desktop: 

```
 cat /etc/apparmor.d/mattermost-desktop
# This profile allows everything and only exists to give the
# application a name instead of having the label "unconfined"

abi <abi/4.0>,
include <tunables/global>

profile mattermost-desktop /usr/bin/mattermost-desktop flags=(unconfined) {
  userns,

# Site-specific additions and overrides. See local/README for details.
  include if exists <local/mattermost-desktop>
}
```

**Comment by theofficialgman** [comment]
> @yvonxiao Could you share links to what exactly the VSCode team did to take care of this? I searched recent pull requests for `apparmor`, `profile`, and `wayland`, but I couldn't find anything relevant.

@ravicious simple answer. vscode did nothing. ubuntu did. ubuntu has vscode (among some other popular 3rd party apps) as part of their `apparmor` profiles https://packages.ubuntu.com/noble/amd64/apparmor/filelist

**Comment by Samueru-sama** [comment]
[This was reported on linuxmint and they rolled back the change that ubuntu did](https://github.com/linuxmint/mint22-beta/issues/82)

Solus did away with apparmor and snap all together instead.

**Comment by Fuseteam** [comment]
There's a discussion on how to handle this issue https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/2046844

**Comment by iongion** [comment]
I am unable to understand this issue, on the bug tracker they mention reverting it, but that is not the case.
What are we supposed to do now besides including a custom apparmor profile for ubuntu debs ?
And if we include it, then what's the point of all this ? ... asking for a friend.
So disappointing and annoying, as if not building open source software to do things / pro bono was not demanding enough.
At least one developer could have tested the effects of this before including it in mainline.

**Comment by hifron** [comment]
The question is if the costs of Windows Apps Store certificate(500 USD) is less than https harrassment or some Snaps policies required for countries problems which flathub currently not assess but is preparing for, so some apparmor profiles harrasment is also in place.

**Comment by Fuseteam** [comment]
It is meant to improve security, and many devs have tested many apps and added fixes for those as seen the discussion, unfortunately those tests were not adequate enough as many slipped through the cracks

**Comment by xmedeko** [comment]
See also feature request to add AppArmor profile to electron-builder request https://github.com/electron-userland/electron-builder/issues/8635

**Comment by jessedegans** [comment]
In the meantime, it would be much better to create a specific AppArmor profile instead of just disabling it system-wide. You can unconfine specific Electron applications like so:

```
abi <abi/4.0>,
include <tunables/global>

profile electron-app /path/to/electron/app flags=(unconfined) {
  userns,
}
```
```
sudo apparmor_parser -r /etc/apparmor.d/electron-app
systemctl --user restart your-electron-app
```

**Comment by xmedeko** [comment]
@jessedegans Is implemented in upcoming electron-builder v26 already, see  https://github.com/electron-userland/electron-builder/issues/8635

**Comment by kdharmay** [comment]
We are on with workaround. 
But, do we have any time line to get the official fix?

**Comment by xmedeko** [comment]
@kdharmay What fix do you mean? If you mean that added AppArmor profile in electron-builder https://github.com/electron-userland/electron-builder/issues/8635, then you have to ask at that project, when electron-builder v26 will be released. (I cannot find any timeline of electron-builder project.)

**Comment by kdharmay** [comment]
> [@kdharmay](https://github.com/kdharmay) What fix do you mean? If you mean that added AppArmor profile in electron-builder [electron-userland/electron-builder#8635](https://github.com/electron-userland/electron-builder/issues/8635), then you have to ask at that project, when electron-builder v26 will be released. (I cannot find any timeline of electron-builder project.)

Thanks @mckravchyk for the response! 
So, the solution would be to create a AppArmor profile to address the issue/crash,
instead of disabling kernel.apparmor_restrict_unprivileged_userns at system level? Please confirm.
Also, may I know why is this bug still open and what is being addressed under this?

**Comment by MikeMcC399** [comment]
There was a new release [electron-builder@26.0.0](https://github.com/electron-userland/electron-builder/releases/tag/v26.0.0) with "feat: add support for AppArmor with template profile and configuration property" which may be helpful. I have not been able to try it personally though.

**Comment by danielweck** [comment]
Annoyingly, ElectronBuilder hasn't released v26 as `latest` NPM package, only `next` which maps with their pre-release GitHub tag:

https://www.npmjs.com/package/electron-builder?activeTab=versions

https://github.com/electron-userland/electron-builder/releases

**Comment by MikeMcC399** [comment]
@danielweck 

I asked the question about the pre-release tag in https://github.com/electron-userland/electron-builder/pull/8808#issuecomment-2618896028.

It is being used to mean that 26.x is ready for production, but not yet widely adopted. Personally, I prefer to hang back and wait until there is a release labeled as `latest`, but it will depend on your situation whether you want to move to 26.x already.

**Comment by xmedeko** [comment]
@danielweck @MikeMcC399 Yes and there's an issue with the same topic already electron-userland/electron-builder#8837 "The 26.0.x versions are labelled as pre-releases on GitHub"

**Comment by danielweck** [comment]
Yep, also note @mmaietta 's statement (core ElectronBuilder maintainer):

https://github.com/electron-userland/electron-builder/issues/8837#issuecomment-2664447412

**Comment by Sytten** [comment]
Bump

**Comment by Fuseteam** [comment]
this should be fixed now tho

**Comment by kdharmay** [comment]
We are currently using the following workaround, as suggested, to mitigate the crash:
sudo sysctl -w kernel.apparmor_restrict_unprivileged_userns=0

Should we instead be creating a custom AppArmor profile (instead of default policy/profile) to properly resolve the crash, rather than disabling the restriction system-wide as shown above?

Could someone please confirm? If that is indeed the correct approach, then what exactly is being addressed by this bug?
CC @jessedegans

**Comment by Fuseteam** [comment]
yes, you just need the proper apparmor profile. My understanding from https://github.com/electron-userland/electron-builder/issues/8635 is that this profile is now generated by the electron builder so developers don't need to write it out themselves

**Comment by kdharmay** [comment]
> yes, you just need the proper apparmor profile. My understanding from [electron-userland/electron-builder#8635](https://github.com/electron-userland/electron-builder/issues/8635) is that this profile is now generated by the electron builder so developers don't need to write it out themselves

Thank you @Fuseteam for the confirmation. 
If we are using cefsimple based application then electron-builder(suggested above) may not be helpful(because Cefsimple is not an electron based application). Given our specific case, I believe we need to write our own AppArmor profile to address the crash.
Please correct me if I am wrong!

**Comment by Fuseteam** [comment]
@kdharmay that is correct, the issue i linked contains a template that you can use and where the profile should go

**Comment by VerteDinde** [comment]
Hey folks, adding to the discussion, I believe this is now something that's handled in Electron Builder & Electron Forge. If you feel this should still be addressed in e/e, let me know and we can look at reopening this issue.

**Comment by Sytten** [comment]
@VerteDinde Both packager does include the profile in deb but for an appimage this is still an issue that needs solving since it create a temporary filesystem which is not always the same so you cant have an apparmor.

**Comment by Methuselah96** [comment]
@VerteDinde Doesn't this still break the ["Building your First App" tutorial](https://www.electronjs.org/docs/latest/tutorial/tutorial-first-app)? Should there be a note on how to deal with this in that tutorial?

**Comment by Fuseteam** [comment]
> [@VerteDinde](https://github.com/VerteDinde) Both packager does include the profile in deb but for an appimage this is still an issue that needs solving since it create a temporary filesystem which is not always the same so you cant have an apparmor.

i think this is best taken up with ubuntu, see https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/2046844 as electrons apps are unlikely the only ones running into this when it comes to appimages
